airbyte-workers/src/main/java/io/airbyte/workers/protocols/airbyte/StateDeltaTracker.java

```diff
@@ -0,0 +1,171 @@
+   * added to this list. Every time a state is committed, state deltas up to the committed state are
+   * removed from the head of the list and aggregated into the committed count map. The source thread
+   * adds while the destination thread removes, so synchronization is necessary to provide
+   * thread-safety.
+   */
+  @VisibleForTesting
+  protected final List<byte[]> stateDeltas;
+
+  @VisibleForTesting
+  protected long remainingCapacity;
+  @VisibleForTesting
+  protected boolean capacityExceeded;
+
+  public StateDeltaTracker(final long memoryLimitBytes) {
+    this.committedStateHashes = new HashSet<>();
+    this.streamToCommittedRecords = new HashMap<>();
+    this.stateDeltas = new ArrayList<>();
+    this.remainingCapacity = memoryLimitBytes;
+    this.capacityExceeded = false;
+  }
+
+  /**
+   * Converts the given state hash and per-stream record count map into a {@code byte[]} and stores
+   * it.
+   * <p>
+   * This method leverages a synchronized block to provide thread safety between the source thread
+   * calling addState while the destination thread calls commitStateHash.
+   *
+   * @throws CapacityExceededException thrown when the memory footprint of stateDeltas exceeds
+   *         available capacity.
+   */
+  public void addState(final int stateHash, final Map<Short, Long> streamIndexToRecordCount) throws CapacityExceededException {
+    final int size = STATE_HASH_BYTES + (streamIndexToRecordCount.size() * BYTES_PER_STREAM);
+
+    if (capacityExceeded || remainingCapacity < size) {
+      capacityExceeded = true;
+      throw new CapacityExceededException("Memory capacity is exceeded for StateDeltaTracker.");
+    }
+
+    final ByteBuffer delta = ByteBuffer.allocate(size);
+
+    delta.putInt(stateHash);
+
+    for (final Map.Entry<Short, Long> entry : streamIndexToRecordCount.entrySet()) {
+      delta.putShort(entry.getKey());
+      delta.putLong(entry.getValue());
+    }
+
+    synchronized (this) {
+      stateDeltas.add(delta.array());
+      remainingCapacity -= delta.array().length;
+    }
+  }
+
+  /**
+   * Mark the given {@code stateHash} as committed.
+   * <p>
+   * This method leverages a synchronized block to provide thread safety between the source thread
+   * calling addState while the destination thread calls commitStateHash.
+   *
+   * @throws StateHashConflictException thrown when the given {@code stateHash} is already committed
+   * @throws CapacityExceededException thrown when committed counts can no longer be reliably computed
+   *         due earlier exceeded capacity from addState.
+   */
+  public void commitStateHash(final int stateHash) throws StateHashConflictException, CapacityExceededException {
+    if (capacityExceeded) {
+      throw new CapacityExceededException("Memory capacity exceeded for StateDeltaTracker, so states cannot be reliably committed");
+    }
+    if (committedStateHashes.contains(stateHash)) {
```

Comment:
i think given how this class / method is called now this is _technically_ safe, but if `commitStateHash` got called from different threads, i think this look up might get funky. i would probably just move your `synchronized` block to include all of the code in this method to avoid anything whacky. 

i tend to err on the side of making thread safety more obviously safe. i think there's essentially no performance cost in this case and if someone is debugging something, they won't do a double take and wonder if an issue is caused by a multi threading problem.